File: src/main/java/org/yearobjectives/domain/entity/util/DateUtils.java
package org.yearobjectives.domain.entity.util;

import java.time.DayOfWeek;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.ZoneOffset;
import java.time.temporal.ChronoUnit;
import java.time.temporal.TemporalAdjusters;


public class DateUtils {

    public static class Day implements MarkerTypeDates {

        @Override
        public Instant getStart() {
            return LocalDateTime.of(LocalDate.now(), LocalTime.MIDNIGHT).toInstant(ZoneOffset.UTC);
        }

        @Override
        public Instant getEnd() {
            return this.getStart()
            .plus(1, ChronoUnit.DAYS);
        }
    }

    public static class Week implements MarkerTypeDates {

        @Override
        public Instant getStart() {
            return LocalDateTime.now().with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY)).withHour(0).withMinute(0).withSecond(0).toInstant(ZoneOffset.UTC);
        }

        @Override
        public Instant getEnd() {
            return this.getStart()
            .plus(7, ChronoUnit.DAYS);
        }
    }

    public static class Month implements MarkerTypeDates {

        @Override
        public Instant getStart() {
            return LocalDateTime.of(LocalDateTime.now().getYear(), LocalDateTime.now().getMonth(), 1, 0, 0, 0).toInstant(ZoneOffset.UTC);
        }

        @Override
        public Instant getEnd() {
            return this.getStart()
            .plus(1, ChronoUnit.MONTHS);
        }
    }

    public static class Year implements MarkerTypeDates {

        @Override
        public Instant getStart() {
            return LocalDateTime.of(LocalDateTime.now().getYear(), 1, 1, 0, 0, 0).toInstant(ZoneOffset.UTC);
        }

        @Override
        public Instant getEnd() {
            return this.getStart()
            .plus(1, ChronoUnit.YEARS);
        }
    }

    private DateUtils() {
    }
}

File: src/main/kotlin/com/objectives/management/domain/entity/utils/DateUtils.kt
package com.objectives.management.domain.entity.utils

import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneOffset
import java.time.temporal.ChronoUnit
import java.time.temporal.TemporalAdjusters

object DateUtils {

    class Day(): DateTypeRange {
        override fun getStartDate(): Instant {
            return LocalDateTime
                .of(LocalDate.now(), LocalTime.MIDNIGHT)
                .toInstant(ZoneOffset.UTC)
        }

        override fun getEndDate(): Instant {
            return getStartDate()
                .plus(1, ChronoUnit.DAYS)
        }
    }

    class Week(): DateTypeRange {
        override fun getStartDate(): Instant {
            return LocalDateTime
                .now()
                .with(TemporalAdjusters
                    .previousOrSame(DayOfWeek.MONDAY))
                .withHour(0)
                .withMinute(0)
                .withSecond(0)
                .toInstant(ZoneOffset.UTC)
        }

        override fun getEndDate(): Instant {
            return getStartDate()
                .plus(7, ChronoUnit.DAYS)
        }
    }

    class Month(): DateTypeRange {
        override fun getStartDate(): Instant {
            return LocalDateTime
                .of(LocalDateTime.now().year, LocalDateTime.now().month, 1, 0, 0, 0)
                .toInstant(ZoneOffset.UTC)
        }

        override fun getEndDate(): Instant {
            return getStartDate()
                .plus(1, ChronoUnit.MONTHS)
        }
    }

    class Year(): DateTypeRange {
        override fun getStartDate(): Instant {
            return LocalDateTime
                .of(LocalDateTime.now().year, 1, 1, 0, 0, 0)
                .toInstant(ZoneOffset.UTC)
        }

        override fun getEndDate(): Instant {
            return getStartDate()
                .plus(1, ChronoUnit.YEARS)
        }
    }

}